course/src/isometry.cc

```diff
@@ -0,0 +1,288 @@
+    if(index > 2) {
+        throw "Invalid index number";
+    }
+    switch(index) {
+        case 0:
+            return x_;
+        case 1:
+            return y_;
+        case 2:
+            return z_;
+    }
+}
+
+Vector3 Vector3::operator + (const Vector3& v) const {
+    return Vector3(x_ + v.x_, y_ + v.y_, z_ + v.z_);
+}
+
+Vector3 Vector3::operator - (const Vector3& v) const {
+    return Vector3(x_ - v.x_, y_ - v.y_, z_ - v.z_);
+}
+
+Vector3 Vector3::operator * (const Vector3& v) const {
+    return Vector3(x_ * v.x_, y_ * v.y_, z_ * v.z_);
+}
+
+Vector3 Vector3::operator * (const double& d) const {
+    return Vector3(x_ * d, y_ * d, z_ * d);
+}
+
+Vector3 Vector3::operator / (const Vector3&v) const {
+    return Vector3(x_ / v.x_, y_ / v.y_, z_ / v.z_);
+}
+
+Vector3& Vector3::operator = (const Vector3& v) {
+    x_ = v.x();
+    y_ = v.y();
+    z_ = v.z();
+    return *this;
+}
+
+Vector3& ekumen::math::Vector3::operator += (const Vector3& v) { 
+    x_ += v.x_;
+    y_ += v.y_;
+    z_ += v.z_;
+    return *this;
+}
+
+Vector3& ekumen::math::Vector3::operator -= (const Vector3& v) {
+    x_ -= v.x_;
+    y_ -= v.y_;
+    z_ -= v.z_;
+    return *this; 
+}
+
+Vector3& ekumen::math::Vector3::operator *= (const Vector3& v) {
+    x_ *= v.x_;
+    y_ *= v.y_;
+    z_ *= v.z_;
+    return *this; 
+}
+
+Vector3& ekumen::math::Vector3::operator /= (const Vector3& v) {
+    x_ /= v.x_;
+    y_ /= v.y_;
+    z_ /= v.z_;
+    return *this; 
+}
+
+bool ekumen::math::Vector3::operator == (const Vector3& v) const {
+    return (x_ == v.x_) && (y_ == v.y_) && (z_ == v.z_);
+}
+
+bool ekumen::math::Vector3::operator != (const Vector3& v) const {
+    return !(*this == v);
+}
+
+bool ekumen::math::Vector3::operator == (const std::initializer_list<double>& list) const {
+    if(list.size() != 3) {
+        throw "Invalid initializer list size";
+    }
+
+    return (*this) == Vector3(list);
+}
+
+bool ekumen::math::Vector3::operator != (const std::initializer_list<double>& list) const {
+    return !(*this == list);
+}
+
+// Computations
+double Vector3::dot(const Vector3& v) const {
+    return (x_ * v.x_) + (y_ * v.y_) + (z_ * v.z_);
+}
+
+double Vector3::norm() const {
+    return sqrt( (x_ * x_) + (y_ * y_) + (z_* z_) );
+}
+
+Vector3 Vector3::cross(const Vector3& v) const {
+    return Vector3(
+    y_ * v.z_ - z_ * v.y_, 
+    z_ * v.x_ - x_ * v.z_,
+    x_ * v.y_ - y_ * v.x_);
+}
+
+// Class constants
+const Vector3 Vector3::kUnitX = Vector3(1,0,0);
+const Vector3 Vector3::kUnitY = Vector3(0,1,0);
+const Vector3 Vector3::kUnitZ = Vector3(0,0,1);
+const Vector3 Vector3::kZero = Vector3(0,0,0);
+
+
+
+
+
+
+// Matrix3 class methods
+
+// Constructors
+Matrix3::Matrix3(const std::initializer_list<double>& list) {
+    if(list.size() != 9) {
+        throw "Invalid initializer list size";
+    }
+    std::initializer_list<double>::iterator it = list.begin();
+    r1_ = Vector3(std::initializer_list<double>{*it++, *it++, *it++});
```

Comment:
Ouch. You've been bitten.

I know that this looks good and "makes sense", but unfortunately the order of evaluation of function arguments C/C++ is undefined. That's fine, unless your function arguments (in this case your iterator auto-increments) have side effects (in this case, they do).

It's impossible to say if the `initializer_list` constructor will compile to be (1st, 2nd, 3rd) as you intended, or (3rd, 2nd, 1st), as a different computer architecture might prefer. 

This is not portable, and it might even not work on a different version of the compiler. 

[See this to get the bad news in more detail](http://www.enseignement.polytechnique.fr/informatique/INF478/docs/Cpp/en/cpp/language/eval_order.html)

Reply:
the devil hides in the details...

Reply:
I suspected that you would comment on this... 
Initially my compiler assigned in the opposite order of the intuitive one (that is, assigning from left to right). So then I added the initializer_list and it worked as expected.
Thanks for this, Gera.

Is there a clean way of doing this without repeating code like the following example?

```
    r1_[0] = *it++;
    r1_[1] = *it++;
    r1_[2] = *it++;

    r2_[0] = *it++;
    r2_[1] = *it++;
    r2_[2] = *it++;

    r3_[0] = *it++;
    r3_[1] = *it++;
    r3_[2] = *it++;
```

I suppose that it would be nice to implement an iterator for Vector3 for following the DRY principle. Right?

Reply:
I'm not sure there's a better way, given that you store the rows in separate variables. I you stored rows in a 3 element array, you could use a pair of nested FOR loops and iterate on them filling all items.

Not sure it's worth changing now, though.